MatchmakingServer/SignalR/QueueingService.cs

```diff
@@ -0,0 +1,875 @@
+        public Player AddPlayer(string connectionId, string playerName)
+        {
+            Player player = new()
+            {
+                Name = playerName,
+                ConnectionId = connectionId
+            };
+
+            if (_connectedPlayers.TryAdd(connectionId, player))
+            {
+                _logger.LogInformation("Connected player: {player}", player);
+            }
+
+            return player;
+        }
+
+        /// <summary>
+        /// Unregister a player.
+        /// </summary>
+        /// <param name="connectionId">Client connection id.</param>
+        /// <returns></returns>
+        public Player? RemovePlayer(string connectionId)
+        {
+            if (!_connectedPlayers.TryRemove(connectionId, out var player))
+            {
+                return null;
+            }
+
+            // clean up
+            DequeuePlayer(player, default, DequeueReason.Disconnect, notifyPlayerDequeued: false);
+
+            return player;
+        }
+
+
+        /// <summary>
+        /// Handle the case when a player reports that the join failed late (e.g. server full) 
+        /// </summary>
+        public void OnPlayerJoinFailed(string connectionId)
+        {
+            if (!_connectedPlayers.TryRemove(connectionId, out var player))
+            {
+                // not found
+                return;
+            }
+
+            if (player.State is not PlayerState.Joining || player.Server is null)
+            {
+                // not joining
+                return;
+            }
+
+            if (player.JoinAttempts.Count >= QueueingSettings.MaxJoinAttempts)
+            {
+                // max join attempts reached -> remove player from queue 
+                DequeuePlayer(player, PlayerState.Connected, DequeueReason.MaxJoinAttemptsReached);
+                return;
+            }
+
+            if (player.Server.LastServerInfo is not null &&
+                player.Server.LastServerInfo.FreeSlots == 0)
+            {
+                // server was probably full
+                // TODO: for this we need to keep polling info
+
+                // allow retry and keep player in queue
+                player.State = PlayerState.Queued;
+                player.Server.JoiningPlayerCount--;
+
+                if (QueueingSettings.ResetJoinAttemptsWhenServerFull)
+                {
+                    // TODO: reset join attempts / timer to make it more fair?
+                    player.JoinAttempts.Clear();
+                }
+
+                return;
+            }
+
+            // allow retry too until timeout or max attempts
+            player.State = PlayerState.Queued;
+            player.Server.JoiningPlayerCount--;
+
+            // otherwise dequeue the player
+            // TODO: maybe allow player to stay in queue until max join attempts / time limit reached
+            //DequeuePlayer(player, PlayerState.Connected, DequeueReason.JoinFailed, notifyPlayerDequeued: true);
```

Comment:
I assume you want to keep this here for the TODO

Reply:
Actually no, i have a really big question mark on how we should handle this. Think we need to collect data and see if we need to change up the failed join handling